src/Features/Core/Portable/CodeRefactoringHelpers.cs

```diff
@@ -3,12 +3,47 @@
 using System.Collections.Immutable;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
 
 namespace Microsoft.CodeAnalysis
 {
     internal static class CodeRefactoringHelpers
     {
+
+        /// <summary>
+        /// <para>
+        /// Returns an instance of <typeparamref name="TSyntaxNode"/> for refactoring given specified selection in document or null
+        /// if no such instance exists.
+        /// </para>
+        /// <para>
+        /// A <typeparamref name="TSyntaxNode"/> instance is returned if selection is zero-width and inside/touching a Token whose
+        /// direct parent is of type <typeparamref name="TSyntaxNode"/> or if a Node of said type is the smallest Node containing
+        /// the whole <paramref name="selection"/>. Otherwise returns <code>null</code>.
+        /// </para>
+        /// <para>
+        /// Note: this function strips all whitespace from both the beginning and the end of given <paramref name="selection"/>.
+        /// The stripped version is then used to determine relevant <see cref="SyntaxNode"/>.
+        /// </para>
+        /// </summary>
+        public static async Task<TSyntaxNode> TryGetSelectedNodeAsync<TSyntaxNode>(
+            Document document, TextSpan selection, CancellationToken cancellationToken, bool traverseUpInSyntaxTreeIfOnEdgeOfSpan = false) where TSyntaxNode : SyntaxNode
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var selectionStripped = await GetStrippedTextSpan(document, selection, cancellationToken).ConfigureAwait(false);
+
+            var node = root.FindNode(selectionStripped) as TSyntaxNode;
+            if (node == null)
+            {
+                // e.g. "C LocalFunction[||](C c)" -> root.FindNode return ParameterList but we still want to return LocalFunctionNode
+                var identifier = await root.SyntaxTree.GetTouchingTokenAsync(selectionStripped.Start,
```

Comment:
you don't know this is an identifier in this helper.

Reply:
I don't quite follow, but the comment is supposed to be just an example showing why this branch is even necessary. A concrete example seemed better to me that to talk generally. (it's not a documentation comment so specificity didn't seem to be an issue if it illustrates point)

Reply:
You wrote `var identifier`, but this is a random helper that may be called on random nodes.  There is no reason at this point to state that this is an identifier.  `var touchingToken` would be reasonable :)

Reply:
Ah I see, the name of the variable. Yep, that should change, copying is frequently the source of problems.